server/src/main/java/de/hellfish/ngrok/server/ServerRunner.java

```diff
@@ -0,0 +1,112 @@
+package de.hellfish.ngrok.server;
+
+
+import lombok.RequiredArgsConstructor;
+import lombok.extern.slf4j.Slf4j;
+import org.springframework.boot.CommandLineRunner;
+import org.springframework.stereotype.Component;
+
+import java.io.*;
+import java.net.ServerSocket;
+import java.net.Socket;
+import java.util.HashMap;
+import java.util.Map;
+import java.util.Optional;
+import java.util.concurrent.ExecutorService;
+import java.util.concurrent.Executors;
+
+@Slf4j
+@Component
+@RequiredArgsConstructor
+public class ServerRunner implements CommandLineRunner {
+
+    private static final int clientToServerPort = 8082;
+    private boolean isRunning = true;
+
+    private final ClientList clientConnections;
+    private final Map<Integer, Socket> userConnections = new HashMap<>();
+    ExecutorService executors = Executors.newFixedThreadPool(5);
+
+
+    @Override
+    public void run(String... args) throws Exception {
+
+        try (ServerSocket clientServerSocket = new ServerSocket(clientToServerPort);
+        ) {
+            log.info("Ngrok-Server started. Listening on clients on port: " + clientToServerPort);
+            Socket clientSocket;
+            while (isRunning) {
+                clientSocket = clientServerSocket.accept();
+                executors.execute(new ClientHandler(clientSocket, clientConnections));
+            }
+        } catch (IOException e) {
+            log.error("Error opening server socket", e);
+        }
+    }
+
+    @RequiredArgsConstructor
+    static class ClientHandler implements Runnable {
+        private final Socket clientSocket;
+        private final ClientList clientList;
+
+        @Override
+        public void run() {
+            log.info("New client connected: " + clientSocket.getLocalAddress() + ":" + clientSocket.getPort());
+            // get init request from client
+            Optional<ClientInitRequest> protocolAndPortOptional = fetchServicePortAndProtocol(clientSocket);
+            if (protocolAndPortOptional.isEmpty()) {
+                log.error("Error on handshake with client");
+                return;
+            }
+            String protocolAndPort = protocolAndPortOptional.get().getValue();
+            // send user-link to client
+            if (clientList.containsProtocolAndPort(protocolAndPort)) {
+                String errorMessage = "ERROR " + protocolAndPort + " is already in use";
+                log.error(errorMessage);
```

Comment:
You don't want to create a separate class where text messages will be stored?

For example:
public static final String MESSAGE = "ERROR %s is already in use";
and then use String.format(MESSAGE, protocolAndPort)

This will make it easier on tests.

Reply:
1:1 I will do it in next pull requests